### PR TITLE
feat: new engine configuration api

### PR DESCRIPTION
**Pull request by darvld**
Ready for review [Powered by Pull Request Badge](https://pullrequestbadge.com/?utm_medium=github&utm_source=elide-dev&utm_campaign=badge_info)

## Summary

> **Important**
> While most features are ready, this PR is a work in progress, and still needs some revision before being considered ready for merge.

This PR implements the new API discussed in #340, which allows for extensive customization of the polyglot runtime using a flexible DSL.
#### Highlights

- Flexible plugin API for customizing GraalVM Engine and Context instances used by the runtime.
- Extensible configuration DSL, available out of the box for all plugins, including custom ones.
- Built-in VFS plugin with support for Host and Embedded guest VFS implementations.
- Basic support for JavaScript, Python, and Ruby, with more in-depth support planned.

```kotlin
fun main() {
  // create and configure aruntime engine
  val engine = PolyglotEngine {
    // add support for languages using plugins
    install(JavaScript) {
      esm = true
      bindings {
        put("message", "Hello, Elide!")
      }
    }

    install(Python) {
      // ...
    }
  }

  // acquire a new context to work with
  val context = engine.acquire()

  // execute guest code via the base APIs
  context.execute(JavaScript, "...")
  
  // or using extensions
  context.javaScript("...")
  context.python("...")
}
```
#### Side effects

- Reflection and DI are no longer used at the runtime layer. The configuration DSL now enables API users to select which features are enabled, and configure them according to their needs. As a consequence, Micronaut is no longer required by the runtime itself (though it will continue to power our CLI and Server modules).
- Concurrency fail-safes such as atomics and locks are no longer needed by the execution engine or the classes providing language support, since configuration options are only available during engine creation.

#### 

#### Opt-in markers

All symbols in the new Core API are annotated with the `@DelicateElideApi` Opt-In marker, which is intended to prevent accidental use of these elements in regular code. Developers working on custom plugins or entry points can safely use the `@OptIn` annotation to remove the compiler warning, as this is the intended use case for these packages.

#### Structure

The new API is divided into packages delimited mainly by language boundaries:

- `runtime-core` provides the core API itself, including the `PolyglotEngine` implementation and the configuration DSL, as well as the Lifecycle and Plugin APIs used for extending the runtime.
- `runtime-js` contains the JavaScript plugin for the runtime and the core intrinsics that are built into it.
- `runtime-py` enables basic support for Python, though at the moment no intrinsics are implemented.
- `runtime-rb` adds support for Ruby, currently in the same state as Python support.

### Host Platform information

A new data class, `HostPlatform`, provides information about the system hosting the runtime, specifically, it lists the Operating System and architecture families.

Engine plugins may access cached host platform data by reading `PolyglotEngineConfiguration.hostPlatform`, which is automatically populated using the `os.name` and `os.arch` JVM System Properties.

### Embedded Resources

Plugin resource manifests have been reworked to allow initialization scripts to be platform-specific (like runtime images already were). Guest VFS layers and "bundles" have been merged into a single list as well.

```kotlin
@Serializable public data class EmbeddedResource(
  val path: String,
  val platform: HostPlatform? = null,
)
```

An `EmbeddedResource` must specify a relative `path` to the target, and _may_ specify a target `platform`, which is expressed in a short lowercase string as `<os>_<arch>` (e.g. `linux_arm64`, `darwin_aarch64`, `windows_amd64`). If no target platform is set, the resource is always loaded.

When loaded using `AbstractLanguagePlugin.resolveEmbeddedManifest()`, an `EmbeddedResource` targeting a different platform than the current one will be filtered out.

### CODE REVIEW

File: tools/scripts/express.js
Comment (sgammon) [reply tools/scripts/express.js:33]: this is from a bad rebase, safe to remove

File: packages/runtime-core/src/main/kotlin/elide/runtime/core/PolyglotEngineConfiguration.kt
Comment (darvld) [inline packages/runtime-core/src/main/kotlin/elide/runtime/core/PolyglotEngineConfiguration.kt:105]: ### Host Platform info

The `PolyglotEngineConfiguration` now exposes cached information about the platform hosting the runtime. This allows plugins to select embedded bundles that match the current platform without handling OS or arch resolution manually.

The `HostPlatform` class and its companion provide functions to resolve host information, and to parse individual components from generic sources (mainly `os.name` and `os.arch` JVM System Properties).

File: packages/runtime-core/src/main/kotlin/elide/runtime/plugins/AbstractLanguagePlugin.kt
Comment (darvld) [inline packages/runtime-core/src/main/kotlin/elide/runtime/plugins/AbstractLanguagePlugin.kt:36]: ### Embedded Resources

Language plugins often use embedded bundles and init scripts that prepare a context for use with the target language. Manifests now use `EmbeddedResource` objects to list both VFS bundles and initialization scripts.

An `EmbeddedResource` must specify a relative `path` to the target, and *may* specify a target `platform`, which is expressed in a short lowercase string as `<os>_<arch>` (e.g. `linux_arm64`, `darwin_aarch64`, `windows_amd64`).

When loaded using `AbstractLanguagePlugin.resolveEmbeddedManifest()`, an `EmbeddedResource` targeting a different platform than the current one will be filtered out.

File: packages/runtime-core/src/main/kotlin/elide/runtime/plugins/AbstractLanguagePlugin.kt
Comment (darvld) [inline packages/runtime-core/src/main/kotlin/elide/runtime/plugins/AbstractLanguagePlugin.kt:35]: ### Embedded Resources

Language plugins often use embedded bundles and init scripts that prepare a context for use with the target language. Manifests now use `EmbeddedResource` objects to list both VFS bundles and initialization scripts.

An `EmbeddedResource` must specify a relative `path` to the target, and *may* specify a target `platform`, which is expressed in a short lowercase string as `<os>_<arch>` (e.g. `linux_arm64`, `darwin_aarch64`, `windows_amd64`).

When loaded using `AbstractLanguagePlugin.resolveEmbeddedManifest()`, an `EmbeddedResource` targeting a different platform than the current one will be filtered out.

File: packages/runtime-core/src/main/kotlin/elide/runtime/core/PolyglotEngineConfiguration.kt
Comment (darvld) [inline packages/runtime-core/src/main/kotlin/elide/runtime/core/PolyglotEngineConfiguration.kt:40]: ### Host Platform

The `HostPlatform` class and its companion provide functions to resolve host information, and to parse individual components from generic sources (mainly `os.name` and `os.arch` JVM System Properties).

Two enums, `OperatingSystem` and `Architecture`, provide well-known values for OS families and architectures, which can then be combined in a single `HostPlatform` record.

File: packages/runtime-js/src/main/kotlin/elide/runtime/plugins/js/Extensions.kt
Comment (sgammon) [inline packages/runtime-js/src/main/kotlin/elide/runtime/plugins/js/Extensions.kt:17]: nit, can it by `javascript`? so it lines up with `python`/`ruby` eventually

File: packages/runtime-js/src/main/resources/META-INF/elide/v4/embedded/runtime/js/js.runtime.tar.gz
Comment (sgammon) [inline packages/runtime-js/src/main/resources/META-INF/elide/v4/embedded/runtime/js/js.runtime.tar.gz:1]: did these get rebuilt? or could these artifact updates be from a rebase?

File: packages/runtime-py/src/main/kotlin/elide/runtime/plugins/python/Extensions.kt
Comment (sgammon) [inline packages/runtime-py/src/main/kotlin/elide/runtime/plugins/python/Extensions.kt:15]: does `/* language=python*/` work!? that's awesome

**Review by sgammon — CHANGES_REQUESTED** [review]
small nits n questions, otherwise, fantastic work 👍🏻

File: packages/cli/build.gradle.kts
Comment (darvld) [reply packages/cli/build.gradle.kts:86]: I don't remember changing this, must be from a bad rebase

File: packages/runtime-js/src/main/kotlin/elide/runtime/plugins/js/Extensions.kt
Comment (darvld) [reply packages/runtime-js/src/main/kotlin/elide/runtime/plugins/js/Extensions.kt:17]: good point, will update

File: packages/runtime-py/src/main/kotlin/elide/runtime/plugins/python/Extensions.kt
Comment (darvld) [reply packages/runtime-py/src/main/kotlin/elide/runtime/plugins/python/Extensions.kt:15]: I might need to verify, I believe it worked with Intellij IDEA

File: packages/runtime-js/src/main/resources/META-INF/elide/v4/embedded/runtime/js/js.runtime.tar.gz
Comment (darvld) [reply packages/runtime-js/src/main/resources/META-INF/elide/v4/embedded/runtime/js/js.runtime.tar.gz:1]: artifacts were placed during development to avoid conflicts with current embedded resources, we might need to discuss the best way of handling it to avoid duplication before merge

File: packages/cli/build.gradle.kts
Comment (sgammon) [reply packages/cli/build.gradle.kts:86]: we'll re-enable on release pr, not a big deal

File: packages/runtime-py/src/main/kotlin/elide/runtime/plugins/python/Extensions.kt
Comment (sgammon) [reply packages/runtime-py/src/main/kotlin/elide/runtime/plugins/python/Extensions.kt:15]: we love to see it

File: packages/runtime-js/src/main/resources/META-INF/elide/v4/embedded/runtime/js/js.runtime.tar.gz
Comment (sgammon) [reply packages/runtime-js/src/main/resources/META-INF/elide/v4/embedded/runtime/js/js.runtime.tar.gz:1]: rgr

**Review by sgammon — APPROVED** [review]
approved pending changes

**Pull request merged**
